src/core/callouts/qgscallout.h

```diff
@@ -0,0 +1,397 @@
+     *
+     * Caller takes ownership of the returned object.
+     */
+    virtual QgsCallout *clone() const = 0 SIP_FACTORY;
+
+    /**
+     * Returns the properties describing the callout encoded in a
+     * string format.
+     *
+     * Subclasses must ensure that they include the base class' properties()
+     * in their returned value.
+     *
+     * \see readProperties()
+     * \see saveProperties()
+     */
+    virtual QVariantMap properties( const QgsReadWriteContext &context ) const;
+
+    /**
+     * Reads a string map of an callout's properties and restores the callout
+     * to the state described by the properties map.
+     *
+     * Subclasses must ensure that they call the base class' readProperties()
+     * method.
+     *
+     * \see properties()
+     */
+    virtual void readProperties( const QVariantMap &props, const QgsReadWriteContext &context );
+
+    /**
+     * Saves the current state of the callout to a DOM \a element. The default
+     * behavior is to save the properties string map returned by
+     * properties().
+     * \returns TRUE if save was successful
+     * \see readProperties()
+     */
+    virtual bool saveProperties( QDomDocument &doc, QDomElement &element, const QgsReadWriteContext &context ) const;
+
+    /**
+     * Restores the callout's properties from a DOM element.
+     *
+     * The default behavior is the read the DOM contents and call readProperties() on the subclass.
+     *
+     * \see readProperties()
+     */
+    virtual void restoreProperties( const QDomElement &element, const QgsReadWriteContext &context );
+
+    /**
+     * Prepares the callout for rendering on the specified render \a context.
+     *
+     * \warning This MUST be called prior to calling render() on the callout, and must always
+     * be accompanied by a corresponding call to stopRender().
+     *
+     * \see stopRender()
+     */
+    virtual void startRender( QgsRenderContext &context );
+
+    /**
+     * Finalises the callout after a set of rendering operations on the specified render \a context.
+     *
+     * \warning This MUST be called after to after render() operations on the callout, and must always
+     * be accompanied by a corresponding prior call to startRender().
+     *
+     * \see startRender()
+     */
+    virtual void stopRender( QgsRenderContext &context );
+
+    /**
+     * Returns the set of attributes referenced by the callout. This includes attributes
+     * required by any data defined properties associated with the callout.
+     *
+     * \warning This must only be called after a corresponding call to startRender() with
+     * the same render \a context.
+     */
+    virtual QSet< QString > referencedFields( const QgsRenderContext &context ) const;
+
+    /**
+     * Renders the callout onto the specified render \a context.
+     *
+     * The \a rect argument gives the desired size and position of the body of the callout (e.g. the
```

Comment:
Shouldn't it be renamed to a more explicit bodyBoundingBox or something ? Just because "rect" is not straightforward

Reply:
Good point, fixed